add accelerate to load models with smaller memory footprint

piEsposito opened the pull request:
Closes #281

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_361). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented on src/diffusers/configuration_utils.py:
```diff
@@ -153,11 +154,17 @@ def from_config(cls, pretrained_model_name_or_path: Union[str, os.PathLike], ret
-        init_dict, unused_kwargs = cls.extract_init_dict(config_dict, **kwargs)
-
-        model = cls(**init_dict)
+        device_map = kwargs.pop("low_cpu_mem_usage", None)
```
Ideally we would like to try to keep `configuration_utils.py` framework and component independent. Could we maybe try to set:

```
with accelerate.init_empty_weights():
            model, unused_kwargs = cls.from_config(
            config_path,
            cache_dir=cache_dir,
            return_unused_kwargs=True,
            force_download=force_download,
            resume_download=resume_download,
            proxies=proxies,
            local_files_only=local_files_only,
            use_auth_token=use_auth_token,
            revision=revision,
            subfolder=subfolder,
            device_map=device_map,
            **kwargs,
        )
```

in `modeling_utils.py` instead?

piEsposito replied:
Just did it.

patrickvonplaten commented on src/diffusers/modeling_utils.py:
```diff
@@ -357,25 +360,41 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
 
-        # Set model in evaluation mode to deactivate DropOut modules by default
-        model.eval()
+        if device_map is not None:            
```
If possible it would be very nice if all `accelerate` logic would only be added here.

piEsposito replied:
Just did it. Had to put the model creation and checkpoint loading after grabbing the weight and config files to avoid splitting the accelerate logic into two.

patrickvonplaten commented on tests/test_models_unet.py:
```diff
@@ -133,6 +133,13 @@ def test_from_pretrained_hub(self):
 
         assert image is not None, "Make sure output is not None"
 
+    def test_from_pretrained_accelerate(self):
```
Cool test! Could we maybe try to also test that:

a) loading with and without accelerate gives the same results when doing a forward pass
b) memory usage is reduced when using accelerate

piEsposito replied:
@patrickvonplaten just did it.

patrickvonplaten reviewed (commented):
Hey @piEsposito,

Sorry for replying only now :-/
Thanks a lot for the PR! It looks already really nice. One important thing that (if possible) would be good to change is only add functionality to `modeling_utils.py` and not `configuration_utils.py` because configuration_utils is also used for the schedulers.

Could you maybe give this a try?

piEsposito commented:
@patrickvonplaten addressed your comments and moved accelerate logic to modelling utils. Also, created some tests to ensure memory usage gets lower and results keep the same. Thank you for your time to carefully review this PR.

patrickvonplaten commented on src/diffusers/modeling_utils.py:
```diff
@@ -20,6 +20,7 @@
 import torch
 from torch import Tensor, device
 
+import accelerate
```
We should not make `accelerate` a hard dependency here . Could you wrap it into a:

```
if accelerate_is_available():
    import accelerate
else:
    raise ImportError("Please install accelerate via `pip install accelerate`")
```

below the `if device_map == "auto"` method?

piEsposito replied:
@patrickvonplaten just did it, thank you for the suggestion.

patrickvonplaten commented on src/diffusers/modeling_utils.py:
```diff
@@ -23,6 +23,7 @@
 from huggingface_hub import hf_hub_download
 from huggingface_hub.utils import EntryNotFoundError, RepositoryNotFoundError, RevisionNotFoundError
 from requests import HTTPError
+from transformers.utils import is_accelerate_available
```
We cannot do this either because `transformers` is not hard requirement :sweat_smile: 

I should have given you more details in my previous comments - very sorry that the feedback cycle takes so much time. Will try hard to reply faster here now. 

In short, can you copy this code https://github.com/huggingface/transformers/blob/e5b7cff5fe65eac9e54ba88fa3935b3270db0207/src/transformers/utils/import_utils.py#L528 into https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/import_utils.py

piEsposito replied:
Hey, thank you for clarifying that. I've just fixed it.

piEsposito commented:
@patrickvonplaten finished implementing all requested changes. Please let me know if anything else comes to your mind.

patrickvonplaten commented:
This looks good to me! 
@patil-suraj @anton-l could you take a look here?

piEsposito commented:
Hey folks, any update here?

patrickvonplaten commented on src/diffusers/modeling_utils.py:
```diff
@@ -293,33 +294,13 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         from_auto_class = kwargs.pop("_from_auto", False)
         torch_dtype = kwargs.pop("torch_dtype", None)
         subfolder = kwargs.pop("subfolder", None)
+        device_map = kwargs.pop("device_map", None)
```
Sorry I oversaw this the first time. 
@piEsposito could you also add some docstring here?

patrickvonplaten replied:
E.g. 3,4 lines under line 264

patrickvonplaten commented:
PR is good to merge for me! Played around with:

```python
#!/usr/bin/env python3
from diffusers import UNet2DConditionModel

model = UNet2DConditionModel.from_pretrained("CompVis/stable-diffusion-v1-3", device_map="auto", subfolder="unet")
import ipdb; ipdb.set_trace()
```

on 1, >1, and no GPU machines and it works as expected.

patrickvonplaten reviewed (approved):
@patil-suraj @anton-l would be nice if one of you could take a look :-)

patrickvonplaten commented:
@piEsposito in a follow-up PR it would be nice if you could implement this then as well to the more global:

```python
from diffusers import DiffusionPipeline

pipeline = DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-3", device_map="auto")
```

functionality :-) Think this could have then very wide-spread adoption. At the moment 99% of users load models via the pipeline interface

patil-suraj commented on tests/test_models_unet.py:
```diff
@@ -133,6 +135,74 @@ def test_from_pretrained_hub(self):
+        model_accelerate.eval()
+
+        noise = torch.randn(
+            1,
+            model_accelerate.config.in_channels,
+            model_accelerate.config.sample_size,
+            model_accelerate.config.sample_size,
+            generator=torch.manual_seed(0),
+        )
+        noise = noise.to(torch_device)
+        time_step = torch.tensor([10] * noise.shape[0]).to(torch_device)
+
+        arr_accelerate = model_accelerate(noise, time_step)["sample"]
+
+        # two models don't need to stay in the device at the same time
+        del model_accelerate
+        torch.cuda.empty_cache()
+        gc.collect()
+
+        model_normal_load, _ = UNet2DModel.from_pretrained("fusing/unet-ldm-dummy-update", output_loading_info=True)
+        model_normal_load.to(torch_device)
+        model_normal_load.eval()
+        arr_normal_load = model_normal_load(noise, time_step)["sample"]
+
+        assert torch.allclose(arr_accelerate, arr_normal_load, rtol=1e-3)
+
+    @unittest.skipIf(torch_device == "cpu", "This test is supposed to run on GPU")
+    def test_memory_footprint_gets_reduced(self):
+        torch.cuda.empty_cache()
+        gc.collect()
+
+        tracemalloc.start()
+        model_accelerate, _ = UNet2DModel.from_pretrained(
+            "fusing/unet-ldm-dummy-update", output_loading_info=True, device_map="auto"
+        )
+        model_accelerate.to(torch_device)
+        model_accelerate.eval()
+        _, peak_accelerate = tracemalloc.get_traced_memory()
+
+        del model_accelerate
```
Very nice tests!

patil-suraj reviewed (approved):
LGTM, thanks a lot for working on this @piEsposito !

piEsposito commented:
> @piEsposito in a follow-up PR it would be nice if you could implement this then as well to the more global:
> 
> 
> 
> ```python
> 
> from diffusers import DiffusionPipeline
> 
> 
> 
> pipeline = DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-3", device_map="auto")
> 
> ```
> 
> 
> 
> functionality :-) Think this could have then very wide-spread adoption. At the moment 99% of users load models via the pipeline interface

@patrickvonplaten great idea. I can start working on that today. Thanks!
